2-advanced-topics/data-pipelines/intro_data_pipelines.ipynb

```diff
@@ -0,0 +1,1037 @@
+    "\n",
+    "**Monitoring and Logging:**\n",
+    "\n",
+    "Implementing monitoring mechanisms to log the pipeline execution, detect failures, and ensure data quality throughout the process."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "ae59b692-f796-434d-90cb-579bd4ad0645",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "![](assets/pipeline-workflow.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "0174d479-d9ef-4ec1-92ab-9b74b2acda2c",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "\n",
+    "The goal of the session is to help you understand how to create basic data pipelines to handle typical data processing tasks and how to modularize these steps to improve maintainability and scalability.\n",
+    "\n",
+    "We will follow a structured approach, starting with data ingestion, then cleaning and transforming the data, and then exporting the processed data and aggregating it to a useful table. Finally, we will use the processed data to produce a plot, to illustrate a typical workflow. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "8759368c-870f-41f0-9f54-3f5000757140",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "## Sample Data (Kenya public expenditure)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "396b6fd8-0201-488a-8740-d3e818522b44",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "For this tutorial we will use the Publicly available Kenya BOOST expenditure dataset available [here]('https://datacatalog.worldbank.org/search/dataset/0038086). This data contains spending information at the central and local levels and continues to be yearly updated, curated and disseminated by the Government of Kenya.\n",
+    "\n",
+    "\n",
+    "We will construct a data pipeline that will ingest this data, make certain modifications, enrich it by adding relevant columns and finally save the resulting tables which would be used to produce some plots to further summarize and understand this expenditure data\n",
+    "\n",
+    "We will use some domain specific understanding of this data, but the process and logic behind the modifications are not domain dependent and illustrate the general principles of data pipeline construction. Transformation operations performed here are typical of the kind of operations encountered when working with raw data"
+   ]
+  },
+  {
```

Comment:
the code part of this block is empty. Make sure to add the code that produces the printed table